File: NET/JNetCall.Sharp.Tests/CallTest.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Example.API;
using JNetBase.Sharp.Sys;
using Microsoft.VisualStudio.TestPlatform.TestHost;
using Xunit;
using static Example.API.ITriggering;

namespace JNetCall.Sharp.Tests
{
    public abstract class CallTest
    {
        protected abstract T Create<T>() where T : class, IDisposable;

        protected virtual string Patch(string input) => input;

        protected virtual int MaxListWait => 1;

        [Fact]
        public void ShouldCallCache()
        {
            var input = new[] { "life", "on", "mars" };
            using var client = Create<IStringCache>();

            client.Set(42, input[0]);
            Assert.Equal(1, client.Size);
            Assert.Equal(input[0], client.Get(42));

            client.Delete(42);
            Assert.Equal(0, client.Size);

            client.Set(43, input[1]);
            client.Set(44, input[2]);
            Assert.Equal(2, client.Size);

            Assert.Equal(input[1], client.Get(43));
            Assert.Equal(input[2], client.Get(44));

            client.Delete(43);
            client.Delete(44);
            Assert.Equal(0, client.Size);

            Assert.Throws<InvalidOperationException>(() => client.Get(89));
            client.Clean();
        }

        [Fact]
        public void ShouldCallMultiple()
        {
            using var client = Create<IMultiple>();

            var t2T = client.GetTuple2T(200, "Greece");
            var t2V = client.GetTuple2V(t2T);
            Assert.Equal(t2T, t2V.ToTuple());

            var t3T = client.GetTuple3T(1, "cat", true);
            var t3V = client.GetTuple3V(t3T);
            Assert.Equal(t3T, t3V.ToTuple());

            var t4T = client.GetTuple4T("perl", new[] { "java", "c#" }, 1, new[] { 2, 3 });
            var t4V = client.GetTuple4V(t4T);
            Assert.Equal(t4T, t4V.ToTuple());

            var t5T = client.GetTuple5T(1, 1.5f, 2L, "dot", "net");
            var t5V = client.GetTuple5V(t5T);
            Assert.Equal(t5T, t5V.ToTuple());

            var bd1 = client.FindBestDay(3);
            Assert.Equal("Wednesday", bd1.ToString());
            var bd2 = client.FindBestDay(5);
            Assert.Equal("Friday", bd2.ToString());
            var bds = client.FindFreeDays();
            Assert.Equal("Sunday, Thursday, Saturday", bds.ToString());

            var bdt = client.GetTextOf(new[] { bd1, bd2 }, bds);
            Assert.Equal("[Wednesday, Friday] | [Sunday, Thursday, Saturday]", bdt);
        }

        [Fact]
        public void ShouldCallDataTyped()
        {
            using var client = Create<IDataTyped>();

            var now = DateTime.Now;
            var dur = TimeSpan.FromSeconds(94);
            var env = new Dictionary<string, int> { { "Caller", nameof(Program).Length } };
            var dict = client.GetSystemVariables(now, dur, env);
            Assert.Equal("[Caller, 7]|[seconds, 94]|[year, 2023]", string.Join("|", dict));

            var lines = new List<string> { "Dog  ", "Hot", "Dog ", "Dog", "Hot    ", "Cat", "Cat", "Hot", "Hot" };
            var lineCount = client.GetLineCount(lines.ToArray());
            var set = client.GetUnique(lines, withTrim: true);
            Assert.Equal("Cat|Dog|Hot", string.Join("|", set));
            Assert.Equal(9, lineCount);

            var list = client.GetDouble(set);
            Assert.Equal("Cat|Dog|Hot|Cat|Dog|Hot", string.Join("|", list));

            var fs = client.AllocateBytes(3, 42);
            Assert.Equal("42|42|42", string.Join("|", fs));

            var bs = client.GetFileSize(@"Z:\Nothing\Good\No fun with that file.txt");
            Assert.Equal(41, bs);

            var a = new
            {
                y = byte.MaxValue, s = short.MaxValue, i = int.MaxValue, l = long.MaxValue, f = float.MaxValue,
                d = double.MaxValue, b = true, c = char.MaxValue, t = "Str", u = decimal.MaxValue, 
                g = Guid.Parse("27edb110-afef-4ce3-b8c1-3fcb2ec3fabe")
            };
            var txt = client.ToSimpleText(a.y, a.s, a.i, a.l, a.f, a.d, a.b, a.c, a.t, a.u, a.g).Replace(" ", "").Trim();
            txt = Patch(txt);
            Assert.Equal("y=-1,s=32767,i=2147483647,l=9223372036854775807,f=3.4028235E38,d=1.7976931348623157E308,b=true,c=￿,t=Str,u=79228162514264337593543950335,g=27edb110-afef-4ce3-b8c1-3fcb2ec3fabe", txt);

            var b = new
            {
                y = new[] { (byte)42 }, s = new[] { short.MinValue }, i = new[] { int.MinValue }, l = new[] { long.MinValue },
                f = new[] { float.MinValue }, d = new[] { double.MinValue }, b = new[] { false }, c = new[] { 'X' },
                t = new[] { "Str1" }, u = new[] { decimal.MinValue }, g = new[] { Guid.Empty }
            };
            txt = client.ToArrayText(b.y, b.s, b.i, b.l, b.f, b.d, b.b, b.c, b.t, b.u, b.g).Replace(" ", "").Trim();
            txt = Patch(txt);
            Assert.Equal("y=[42],s=[-32768],i=[-2147483648],l=[-9223372036854775808],f=[-3.4028235E38],d=[-1.7976931348623157E308],b=[false],c=[X],t=[Str1],u=[-79228162514264337593543950335],g=[00000000-0000-0000-0000-000000000000]", txt);
        }

        [Fact]
        public async Task ShouldCallSimultan()
        {
            using var client = Create<ISimultaneous>();

            await client.LoadIt("Hello");

            var id = await client.GetId();
            Assert.True(id is >= -100 and <= 100, id + " ?!");

            var txt = await client.RemoveIt();
            Assert.Equal("Hello", txt);

            const int count = 3;
            var range = Enumerable.Range(0, count);

            var watch = Stopwatch.StartNew();
            var tasks = range.Select(i => client.RunIt(26, i))
                .ToList();
            watch.Stop();
            var listTime = watch.ElapsedMilliseconds;

            var all = Task.WhenAll(tasks).GetAwaiter().GetResult();
            Assert.Equal(count, all.Length);

            var numbers = all.Select(t => t.Item1).Distinct().ToArray();
            Assert.Equal(count, numbers.Length);

            Assert.True(listTime >= 0 && listTime <= MaxListWait, listTime + " ?!");
        }

        [Fact]
        public void ShouldCallCalculator()
        {
            using var client = Create<ICalculator>();
            Assert.Equal("Java", client.Name);

            var value1 = 100.00D;
            var value2 = 15.99D;
            var result = client.Add(value1, value2);
            Assert.Equal(115.99, result);

            value1 = 145.00D;
            value2 = 76.54D;
            result = client.Subtract(value1, value2);
            Assert.Equal(68.46, result);

            value1 = 9.00D;
            value2 = 81.25D;
            result = client.Multiply(value1, value2);
            Assert.Equal(731.25, result);

            value1 = 22.00D;
            value2 = 7.00D;
            result = client.Divide(value1, value2);
            Assert.Equal(3.142857142857143, result);
        }

        [Fact]
        public void ShouldCallTrigger()
        {
            using var client = Create<ITriggering>();

            const int cbCount = 3;
            var cbList = new List<Tuple<int, string>>();
            var clc = new CountdownEvent(cbCount);

            bool EnumWindowsCallback(int handle, string lParam)
            {
                cbList.Add(Tuple.Create(handle, lParam));
                clc.Signal();
                return true;
            }

            var callOk = client.EnumWindows(EnumWindowsCallback, cbCount);
            Assert.True(callOk);
            clc.Wait(TimeSpan.FromSeconds(5));

            Assert.Equal(cbCount, cbList.Count);
            Assert.Equal("[(0, 3!), (1, 4!), (2, 5!)]", Arrays.ToString(cbList));

            const int evtCount = 4;
            var evtList = new List<Tuple<string, int>>();
            var cle = new CountdownEvent(evtCount);

            void OnThresholdReached(object sender, ThresholdEventArgs e)
            {
                var s = sender.ToString();
                evtList.Add(Tuple.Create(s, e.Threshold));
                cle.Signal();
            }

            client.ThresholdReached += OnThresholdReached;
            client.StartPub(evtCount);
            cle.Wait(TimeSpan.FromSeconds(5));

            Assert.Equal(evtCount, evtList.Count);
            client.ThresholdReached -= OnThresholdReached;
        }
    }
}

File: JVM/jnetcall-java-tests/src/test/java/jnetcall/java/tests/CallTest.java
package jnetcall.java.tests;

import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertThrows;
import static org.testng.Assert.assertTrue;

import java.math.BigDecimal;
import java.time.Duration;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.stream.IntStream;

import org.example.api.ICalculator;
import org.example.api.IDataTyped;
import org.example.api.IMultiple;
import org.example.api.IMultiple.Days;
import org.example.api.IMultiple.WeekDay;
import org.example.api.ISimultaneous;
import org.example.api.IStringCache;
import org.example.api.ITriggering;
import org.example.api.ITriggering.PCallBack;
import org.example.api.ITriggering.ThresholdHandler;
import org.javatuples.Pair;
import org.javatuples.Quartet;
import org.javatuples.Quintet;
import org.javatuples.Triplet;
import org.testng.annotations.Test;
import org.testng.util.Strings;

import com.google.common.base.Stopwatch;
import com.xafero.javaenums.BitFlag;

import jnetbase.java.sys.Primitives;
import jnetbase.java.threads.Tasks;

public abstract class CallTest {

    protected abstract <T extends AutoCloseable> T create(Class<T> clazz);

    protected String patch(String input) { return input; }

    protected int getMaxListWait() { return 1; }

    @Test
    public void shouldCallCache() throws Exception {
        String[] input = new String[]{"life", "on", "mars"};
        try (IStringCache client = create(IStringCache.class)) {

            client.set(42, input[0]);
            assertEquals(1, client.getSize());
            assertEquals(input[0], client.get(42));

            client.delete(42);
            assertEquals(0, client.getSize());

            client.set(43, input[1]);
            client.set(44, input[2]);
            assertEquals(2, client.getSize());

            assertEquals(input[1], client.get(43));
            assertEquals(input[2], client.get(44));

            client.delete(43);
            client.delete(44);
            assertEquals(0, client.getSize());

            assertThrows(UnsupportedOperationException.class, () -> client.get(89));
            client.clean();
        }
    }

    @Test
    public void shouldCallMultiple() throws Exception {
        try (IMultiple client = create(IMultiple.class)) {

            Pair<Integer, String> t2T = client.GetTuple2T(200, "Greece");
            Pair<Integer, String> t2V = client.GetTuple2V(t2T);
            assertEquals(t2T, t2V);

            Triplet<Integer, String, Boolean> t3T = client.GetTuple3T(1, "cat", true);
            Triplet<Integer, String, Boolean> t3V = client.GetTuple3V(t3T);
            assertEquals(t3T, t3V);

            Quartet<String, String[], Integer, int[]> t4T = client.GetTuple4T("perl", new String[]{"java", "c#"}, 1, new int[]{2, 3});
            Quartet<String, String[], Integer, int[]> t4V = client.GetTuple4V(t4T);
            assertEquals(t4T.getValue0(), t4V.getValue0());
            assertEquals(t4T.getValue1(), t4V.getValue1());
            assertEquals(t4T.getValue2(), t4V.getValue2());
            assertEquals(t4T.getValue3(), t4V.getValue3());

            Quintet<Integer, Float, Long, String, String> t5T = client.GetTuple5T(1, 1.5f, 2L, "dot", "net");
            Quintet<Integer, Float, Long, String, String> t5V = client.GetTuple5V(t5T);
            assertEquals(t5T, t5V);

            WeekDay bd1 = client.FindBestDay(3);
            assertEquals("Wednesday", bd1.toString());
            WeekDay bd2 = client.FindBestDay(5);
            assertEquals("Friday", bd2.toString());
            BitFlag<Days> bds = client.FindFreeDays();
            assertEquals("[Sunday, Thursday, Saturday]", bds.toString());

            String bdt = client.GetTextOf(new IMultiple.WeekDay[]{bd1, bd2}, bds);
            assertEquals("[Wednesday, Friday] | [Sunday, Thursday, Saturday]", bdt);
        }
    }

    @Test
    public void shouldCallDataTyped() throws Exception {
        try (IDataTyped client = create(IDataTyped.class)) {

            LocalDateTime now = LocalDateTime.now();
            Duration dur = Duration.ofSeconds(94);
            HashMap<String, Integer> env = new HashMap<String, Integer>();
            env.put("Caller", "Program".length());
            String[] dict = client.GetSystemVariables(now, dur, env)
                    .entrySet().stream().map(f -> '['+f.getKey()+", "+f.getValue()+']').toArray(String[]::new);
            assertEquals("[Caller, 7]|[seconds, 94]|[year, 2023]", Strings.join("|", dict));

            List<String> lines = Arrays.asList(new String[]{"Dog  ", "Hot", "Dog ", "Dog", "Hot    ", "Cat", "Cat", "Hot", "Hot"});
            int lineCount = client.GetLineCount(lines.toArray(String[]::new));
            Set<String> set = client.GetUnique(lines, true);
            assertEquals("Cat|Dog|Hot", Strings.join("|", set.stream().toArray(String[]::new)));
            assertEquals(9, lineCount);

            String[] list = client.GetDouble(set).toArray(String[]::new);
            assertEquals("Cat|Dog|Hot|Cat|Dog|Hot", Strings.join("|", list));

            String[] fs = Arrays.stream(Primitives.castInt(client.AllocateBytes(3, (byte) 42)))
                    .boxed().map(f -> f.toString()).toArray(String[]::new);
            assertEquals("42|42|42", Strings.join("|", fs));

            long bsf = client.GetFileSize("Z:\\Nothing\\Good\\No fun with that file.txt");
            assertEquals(41, bsf);

            byte ay = Byte.MAX_VALUE; short as = Short.MAX_VALUE; int ai = Integer.MAX_VALUE;
            long al = Long.MAX_VALUE; float af = Float.MAX_VALUE; double ad = Double.MAX_VALUE;
            boolean ab = true; char ac = Character.MAX_VALUE; String at = "Str";
            BigDecimal au = BigDecimal.valueOf(Long.MAX_VALUE);
            UUID ag = UUID.fromString("27edb110-afef-4ce3-b8c1-3fcb2ec3fabe");
            String txt = client.ToSimpleText(ay, as, ai, al, af, ad, ab, ac, at, au, ag).replace(" ", "")
                    .trim().replace("3,4", "3.4").replace("1,7", "1.7");
            txt = patch(txt);
            assertEquals("y=127,s=32767,i=2147483647,l=9223372036854775807,f=3.4028235E+38,d=1.7976931348623157E+308,b=True,c=￿,t=Str,u=9223372036854775807,g=27edb110-afef-4ce3-b8c1-3fcb2ec3fabe", txt);

            byte[] by = new byte[]{ 42 }; short[] bs = new short[]{ Short.MIN_VALUE }; int[] bi = new int[]{ Integer.MIN_VALUE };
            long[] bl = new long[]{ Long.MIN_VALUE }; float[] bf = new float[]{ -3.4028235E38f };
            double[] bd = new double[]{ -1.7976931348623157E308d }; boolean[] bb = new boolean[]{ false };
            char[] bc = new char[]{ 'X' }; String[] bt = new String[]{ "Str1" };
            BigDecimal[] bu = new BigDecimal[] { BigDecimal.valueOf(Long.MIN_VALUE) };
            UUID[] bg = new UUID[]{ UUID.fromString("00000000-0000-0000-0000-000000000000") };
            txt = client.ToArrayText(by, bs, bi, bl, bf, bd, bb, bc, bt, bu, bg).replace(" ", "")
                    .trim().replace("3,4", "3.4").replace("1,7", "1.7");
            txt = patch(txt);
            assertEquals("y=[42],s=[-32768],i=[-2147483648],l=[-9223372036854775808],f=[-3.4028235E+38],d=[-1.7976931348623157E+308],b=[False],c=[X],t=[Str1],u=[-9223372036854775808],g=[00000000-0000-0000-0000-000000000000]", txt);
        }
    }

    @Test
    public void shouldCallSimultan() throws Exception {
        try (ISimultaneous client = create(ISimultaneous.class)) {

            client.loadIt("Hello").get();

            Integer id = client.getId().get();
            assertTrue(id >= -100 && id <= 100, id + " ?!");

            String txt = client.removeIt().get();
            assertEquals("Hello", txt);

            final int count = 3;
            IntStream range = IntStream.range(0, count);

            Stopwatch watch = Stopwatch.createStarted();
            List<CompletableFuture<Pair<Integer, Long>>> tasks = range.mapToObj(i -> client.runIt(26, i)).toList();
            watch.stop();
            long listTime = watch.elapsed(TimeUnit.MILLISECONDS);

            List<Pair<Integer, Long>> all = Tasks.whenAll(tasks);
            assertEquals(count, all.size());

            Object[] numbers = all.stream().map(t -> t.getValue0()).distinct().toArray();
            assertEquals(count, numbers.length);

            assertTrue(listTime >= 0 && listTime <= getMaxListWait(), listTime + " ?!");
        }
    }

    @Test
    public void shouldCallCalculator() throws Exception {
        try (ICalculator client = create(ICalculator.class)) {

            assertEquals("C#", client.getName());

            double value1 = 100.00D;
            double value2 = 15.99D;
            double result = client.add(value1, value2);
            assertEquals(115.99, result);

            value1 = 145.00D;
            value2 = 76.54D;
            result = client.subtract(value1, value2);
            assertEquals(68.46, result);

            value1 = 9.00D;
            value2 = 81.25D;
            result = client.multiply(value1, value2);
            assertEquals(731.25, result);

            value1 = 22.00D;
            value2 = 7.00D;
            result = client.divide(value1, value2);
            assertEquals(3.142857142857143, result);
        }
    }

    @Test
    public void shouldCallTrigger() throws Exception {
        try (ITriggering client = create(ITriggering.class)) {

            final int cbCount = 3;
            ArrayList<Pair<Integer, String>> cbList = new ArrayList<Pair<Integer, String>>();
            final CountDownLatch clc = new CountDownLatch(cbCount);

            PCallBack EnumWindowsCallback = (handle, lParam) -> {
                cbList.add(Pair.with(handle, lParam));
                clc.countDown();
                return true;
            };

            boolean callOk = client.enumWindows(EnumWindowsCallback, cbCount);
            assertTrue(callOk);
            clc.await(5, TimeUnit.SECONDS);

            assertEquals(cbCount, cbList.size());
            assertEquals("[[0, 3!], [1, 4!], [2, 5!]]", cbList.toString());

            final int evtCount = 4;
            ArrayList<Pair<String, Integer>> evtList = new ArrayList<Pair<String, Integer>>();
            final CountDownLatch cle = new CountDownLatch(evtCount);

            ThresholdHandler OnThresholdReached = (sender, e) -> {
                String s = sender.toString();
                evtList.add(Pair.with(s, e.Threshold()));
                cle.countDown();
            };

            client.addThresholdReached(OnThresholdReached);
            client.startPub(evtCount);
            cle.await(5, TimeUnit.SECONDS);

            assertEquals(evtCount, evtList.size());
            client.removeThresholdReached(OnThresholdReached);
        }
    }
}
